napari/layers/labels/labels.py

```diff
@@ -525,7 +537,37 @@ def paint(self, coord, new_label, refresh=True):
             slice_coord = tuple(slice_coord)
 
         # update the labels image
-        self.data[slice_coord] = new_label
+
+        if self._overwrite:
+            self.data[slice_coord] = new_label
+        else:
+            self.paint_without_overwrite(self.data, slice_coord, new_label, 0)
```

Comment:
I wonder if instead of having a new method here which contains for loops if there is a faster method for grabbing the coordinates of the non-background values of the numpy array and then just writing into those directly - @jni or @kne42 might have an idea here.

Reply:
Maybe something like
```python
np.nonzero(self.data[slice_coord] == self.background_label)
```
though you'll have to make the coords to this new object line up with the coords of the full object. Generally though approaches using numpy commands like this tend to be faster I think.

Reply:
I am a bit unsure about this as well, some expertise would be much appreciated. The challenge I am trying to work around is that slice hides each individual coordinate, where we would need an explicit mapped out matrix to track status for each point.

Reply:
I just checked if you just do
```python
keep_coords = np.nonzero(self.data[slice_coord] == self._background_label)
self.data[slice_coord][keep_coords] = new_label
```
then everything works great - no need for any additional functions

Reply:
Good stuff! We can do one simpler:

```python
keep_coords = (self.data[slice_coord] == self._background_label)
```

Reply:
huuray!! thanks for the super cool one liner! 
on a related topic, I am bootstraping myself to python (I am able to write somewhat functional but lengthy code like this, but not able to produce elegant solutions like what you suggested) any recommended reading on the nice python tips/tricks?

Reply:
> I am bootstraping myself to python (I am able to write somewhat functional but lengthy code like this, but not able to produce elegant solutions like what you suggested) any recommended reading on the nice python tips/tricks?

I'm going to give a shout out to [Elegant SciPy: The Art Scientific Python](https://www.amazon.com/Elegant-SciPy-Art-Scientific-Python/dp/1491922877) written by our very own @jni! I should actually get myself a copy too now that I think about it ;-)

Also available as a repo here https://github.com/elegant-scipy/elegant-scipy

Reply:
> I'm going to give a shout out to Elegant SciPy: The Art Scientific Python

It's a good book. =P Another famous book in the space is the Python Data Science Handbook by Jake VanderPlas: https://jakevdp.github.io/PythonDataScienceHandbook/

Finally, not everything in @stefanv's and my Advanced NumPy tutorial made it onto the book, so that's a good place to get good ideas on what's possible with NumPy indexing:

https://github.com/ASPP/2019-camerino-advanced-numpy/blob/master/Advanced%20NumPy%20Patterns.ipynb